lib/rules/semi-style.js

```diff
@@ -25,6 +25,82 @@ const SELECTOR = `:matches(${
     ].join(",")
 })`;
 
+/**
+ * Get the child node list of a given node.
+ * This returns `Program#body`, `BlockStatement#body`, or `SwitchCase#consequent`.
+ * This is used to check whether a node is the first/last child.
+ * @param {Node} node A node to get child node list.
+ * @returns {Node[]|null} The child node list.
+ */
+function getChildren(node) {
+    const t = node.type;
+
+    if (t === "BlockStatement" || t === "Program") {
+        return node.body;
+    }
+    if (t === "SwitchCase") {
+        return node.consequent;
+    }
+    return null;
+}
+
+/**
+ * Check whether a given node is the first statement in the parent block.
+ * @param {Node} node A node to check.
+ * @returns {boolean} `true` if the node is the first statement in the parent block.
+ */
+function isFirstChild(node) {
+    const nodeList = getChildren(node.parent);
+
+    return nodeList !== null && nodeList[0] === node;
+}
+
+/**
+ * Check whether a given node is the last statement in the parent block.
+ * @param {Node} node A node to check.
+ * @returns {boolean} `true` if the node is the last statement in the parent block.
+ */
+function isLastChild(node) {
+    const t = node.parent.type;
+
+    if (t === "IfStatement" && node.parent.consequent === node && node.parent.alternate) { // before `else` keyword.
+        return true;
+    }
+    if (t === "DoWhileStatement") { // before `while` keyword.
+        return true;
+    }
+    const nodeList = getChildren(node.parent);
+
+    return nodeList !== null && nodeList[nodeList.length - 1] === node; // before `}` or etc.
+}
+
+/**
+ * Check whether a given node is a dangling semi. For example:
+ *
+ *     while (a)
+ *         ;
+ *
+ * @param {Node} node A node to check.
+ * @returns {boolean} `true` if the node is the last statement in the parent block.
+ */
+function isDanglingEmptyStatement(node) {
+    if (node.type !== "EmptyStatement") {
+        return false;
+    }
+    const t = node.parent.type;
+
+    return (
+        t === "DoWhileStatement" ||
+        t === "ForInStatement" ||
+        t === "ForOfStatement" ||
+        t === "ForStatement" ||
+        t === "IfStatement" ||
+        t === "LabeledStatement" ||
+        t === "WhileStatement" ||
+        t === "WithStatement"
```

Comment:
Could this be `!astUtils.STATEMENT_LIST_PARENTS.has(node.parent.type)`?

Reply:
Oh, that's simple. Yes, it works fine!

Reply:
After I rethought, I'm wondering if I remove the check of `EmptyStatement` entirely. It makes only redundant consequent semicolons... E.g. `foo;;\nbar` or `foo\n;;bar`. `No-extra-semi` rule should remove that.